src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+
+#include "i18n.h"
+
+#include <workerd/jsg/exception.h>
+
+#include <unicode/putil.h>
+#include <unicode/timezone.h>
+#include <unicode/uchar.h>
+#include <unicode/uclean.h>
+#include <unicode/ucnv.h>
+#include <unicode/udata.h>
+#include <unicode/uidna.h>
+#include <unicode/ulocdata.h>
+#include <unicode/urename.h>
+#include <unicode/ustring.h>
+#include <unicode/utf16.h>
+#include <unicode/utf8.h>
+#include <unicode/utypes.h>
+#include <unicode/uvernum.h>
+#include <unicode/uversion.h>
+
+#include "simdutf.h"
+
+namespace workerd::api::node {
+
+namespace i18n {
+
+namespace {
+
+// An isolate has a 128mb memory limit.
+const int ISOLATE_LIMIT = 134217728;
+
+struct ConverterDisposer : public kj::Disposer {
+  static const ConverterDisposer INSTANCE;
+  void disposeImpl(void* pointer) const override {
+    ucnv_close(reinterpret_cast<UConverter*>(pointer));
+  }
+};
+
+const ConverterDisposer ConverterDisposer::INSTANCE;
+
+const char* getEncodingName(Encoding input) {
+  switch (input) {
+  case Encoding::ASCII:
+    return "us-ascii";
+  case Encoding::LATIN1:
+    return "iso8859-1";
+  case Encoding::UTF16LE:
+    return "utf16le";
+  case Encoding::UTF8:
+    return "utf-8";
+  default:
+    KJ_UNREACHABLE;
+  }
+}
+
+typedef kj::Maybe<kj::Array<kj::byte>> (*TranscodeImpl)(kj::ArrayPtr<kj::byte> source,
+                                                        Encoding fromEncoding, Encoding toEncoding);
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeDefault(kj::ArrayPtr<kj::byte> source,
+                                                Encoding fromEncoding, Encoding toEncoding) {
+  Converter to(toEncoding);
+  auto substitude = kj::str(kj::repeat('?', to.minCharSize()));
+  to.setSubstitudeChars(substitude);
+  Converter from(fromEncoding);
+
+  size_t limit = source.size() + to.maxCharSize();
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(limit <= ISOLATE_LIMIT, Error, "Source buffer is too large to transcode");
+  auto out = kj::heapArray<kj::byte>(limit);
+  char* target = out.asChars().begin();
+  const char* source_ = source.asChars().begin();
+  UErrorCode status{};
+  ucnv_convertEx(to.conv(), from.conv(), &target, target + limit, &source_, source_ + source.size(),
+                 nullptr, nullptr, nullptr, nullptr, true, true, &status);
+  if (U_SUCCESS(status)) {
+    return out.slice(0, target - out.asChars().begin()).attach(kj::mv(out));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeToUCS2(kj::ArrayPtr<kj::byte> source, Encoding fromEncoding,
```

Comment:
If this function only transcodes from latin1 I think it's better to rename it and remove the fromencoding arg